### ISSUE TITLE
Initial set of ops for MNIST training work e2e

### ISSUE THREAD

**Issue by nvukobratTT** [tenstorrent/tt-mlir#145]
### Summary
Key goal is to have e2e tests that can cover each op. E.g. to support the following flow:
```
TVM -> PyBuda Compile -> MLIR -> Flattbuffer -> PyBuda Runtime
```

### Details
Unique op types that are required for initial support are:
- add
- subtract
- multiply
- matmul
- reduce sum
- softmax

Some of the op tests already have their placeholders. More details on this path: `pybuda/test/mlir/test_ops.py`. Depending on the runtime API changes, `torch.compile` should be replaced with `pybuda.compile`.

Also, here is the list of ops required for the current implementation of the MNIST based Linear model, split across different compotents:
- Forward (includes loss): add, matmul, softmax (not decomposed)
- Backward: add, matmul, multiply, softmax_bwd (decomposed; multiply, reduce_sum, subtract)
- Optimizer: multiply, subtract

Majority of ops already have MLIR support. We need to map those and lower to the MLIR IR accordingly.

### Passing criteria
- All defined ops work e2e with matching golden/silicon PCC

**Comment by vladimirjovanovicTT** [comment]
We need to doublecheck this list for the current state of frontend.
We also need to make sure that training and inference use a nice MNIST implementation (current MNIST model used for inference test has a few workarounds).

**Comment by sdjordjevicTT** [comment]
@vladimirjovanovicTT: I created the following issue to rewrite the MNIST model: https://github.com/tenstorrent/tt-forge-fe/issues/164

@nvukobratTT: @vladimirjovanovicTT, do you guys see anything wrong with moving this issue to the tt-forge-fe repo? This issue mostly captures the e2e forge scenario, so I don't see the point of it being present here in the tt-mlir repo. For this repo, I would like to see the specific ops that need to be implemented, rather than having an e2e scenario. Let me know what you guys think about it.

**Comment by nvukobratTT** [comment]
> @vladimirjovanovicTT: I created the following issue to rewrite the MNIST model: [tenstorrent/tt-forge-fe#164](https://github.com/tenstorrent/tt-forge-fe/issues/164)
> 
> @nvukobratTT: @vladimirjovanovicTT, do you guys see anything wrong with moving this issue to the tt-forge-fe repo? This issue mostly captures the e2e forge scenario, so I don't see the point of it being present here in the tt-mlir repo. For this repo, I would like to see the specific ops that need to be implemented, rather than having an e2e scenario. Let me know what you guys think about it.

Most of these issues should be transferred to the FFE repo. @vladimirjovanovicTT can you push this further in the following days? :))

**Comment by vladimirjovanovicTT** [comment]
Hopefully we'll move them today to FFE.

**Comment by sdjordjevicTT** [comment]
Great! Thanks, folks!
